### PR TITLE
Translate generated pgsql error locations to input text

### PR DESCRIPTION
**Pull request by efaust**
When a user writes some error-ful query into psql, we are currently returning
errors in terms of the generated query. This is unsightly and hard to follow:

```
edgedb=# select 1 + 'goo';
ERROR:  invalid input syntax for type integer: "goo"
LINE 2:     (1 + 'goo')
                 ^
QUERY:  SELECT/*<pg.SelectStmt at 0x1213e37c0>*/
    (1 + 'goo')
```

Instead, return error locations against the input query. This yields:

```
edgedb=# select 1 + 'goo';
ERROR:  invalid input syntax for type integer: "goo"
LINE 1: select 1 + 'goo';
                   ^
QUERY:  select 1 + 'goo';
```

Much better.

### CODE REVIEW

File: edb/common/ast/base.py [hunk 1]
```diff
@@ -243,7 +243,10 @@ def __init__(self, **kwargs):
     def __copy__(self):
         copied = self.__class__()
         for field, value in iter_fields(self, include_meta=False):
-            object.__setattr__(copied, field, value)
+            try:
```
Comment (efaust) [inline edb/common/ast/base.py:246]: Some fields are readonly, and we hit this in the impl of `.replace()`

Pretty scary/ugly to skip here.

File: edb/pgsql/ast.py [hunk 2]
```diff
@@ -667,10 +667,7 @@ class Expr(ImmutableBaseExpr):
 
 
 class BaseConstant(ImmutableBaseExpr):
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
-        if not isinstance(self, NullConstant) and self.val is None:
-            raise ValueError('cannot create a pgast.Constant without a value')
+    pass
```
Comment (efaust) [inline edb/pgsql/ast.py:670]: Again, `.replace()` calls __init__(), and then copies the fields across, so we lose our pretty safety check.

File: edb/pgsql/codegen.py [hunk 3]
```diff
@@ -117,6 +145,57 @@ def ctes_to_source(cls, ctes: List[pgast.CommonTableExpr]) -> str:
...
+    ) -> None:
+        start = len(self.result)
+        super().write(*x, delimiter=delimiter)
+        for new in range(start, len(self.result)):
+            self.write_index += len(self.result[new])
+
+    def visit(self, node):  # type: ignore
+        if self.with_translation_data:
+            translation_data = SQLSourceGeneratorTranslationData(
+                source_start=node.context.start if node.context else 0,
```
Comment (efaust) [inline edb/pgsql/codegen.py:186]: This needs a comment. Only the outermost statement lacks a context.

File: edb/common/ast/codegen.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def visit_indented(
         if indent:
             self.new_lines = 1
             self.char_indentation += 1
-        res = super().visit(node)
+        res = self.visit(node)
```
Comment (efaust) [inline edb/common/ast/codegen.py:63]: _please_ someone check this. This looks right to me, but it also scares the bejeezus out of me.

File: edb/common/ast/base.py (hunk 1 above)
Comment (msullivan) [reply edb/common/ast/base.py:246]: Add a comment

File: edb/server/compiler/compiler.py [hunk 5]
```diff
@@ -668,8 +668,14 @@ def parse_search_path(search_path_str: str) -> list[str]:
                     **args
                 )
                 resolved = pg_resolver.resolve(stmt, schema, options)
-                source = pg_codegen.generate_source(resolved)
-                unit = dbstate.SQLQueryUnit(query=source)
+                source, tl_data = \
```
Comment (msullivan) [inline edb/server/compiler/compiler.py:671]: Use parens instead of a backslash. I hate backslashes

File: edb/common/ast/codegen.py (hunk 4 above)
Comment (msullivan) [reply edb/common/ast/codegen.py:63]: This seems like it should be right, to me.

But also some tests are failing with syntax errors so maybe we're wrong

**Review by msullivan — COMMENTED** [review]
Looks pretty good!

Needs tests, and it seems to have broken some existing tests legitimately

File: edb/server/pgcon/pgcon.pyx [hunk 6]
```diff
@@ -1667,13 +1667,21 @@ cdef class PGConnection:
                             field_type = self.buffer.read_byte()
                             if field_type == b'P':  # Position
                                 msg_buf.write_byte(b'q')  # Internal query
-                                msg_buf.write_bytestring(query)
+                                msg_buf.write_bytestring(unit.input_query.encode('utf8'))
```
Comment (fantix) [inline edb/server/pgcon/pgcon.pyx:1670]: After translation, we don't need to include the original query with "internal query" (`q`) any more - we can simply return the translated data with the "position" (`P`), instead of the "internal position" (`p`). I believe psql already has the SQL and it could produce a proper error output.

**Review by fantix — APPROVED** [review]
Other than the issue above, I think this is good!

**Review by msullivan — APPROVED** [review]
Looks good to me! I think fantix is right about position v. internal position, so follow up on that. (Which might need a test change.)

**Comment by efaust** [comment]
Updated for extended queries. Would like @fantix to take another look to make sure I'm not crazy

File: edb/server/pgcon/pgcon.pyx [hunk 7]
```diff
@@ -2062,12 +2086,17 @@ cdef class PGConnection:
...
                 if field_type == b'P':  # Position
-                    # Internal query
-                    msg_buf.write_byte(b'q')
-                    # Compiled SQL
-                    msg_buf.write_bytestring(action.args[0])
-                    # Internal position
-                    msg_buf.write_byte(b'p')
+                    qu = (action.query_unit.translation_data
+                          if isinstance(action.query_unit,
+                                        compiler.dbstate.SQLQueryUnit)
```
Comment (fantix) [inline edb/server/pgcon/pgcon.pyx:2091]: `action.query_unit` has type `SQLQueryUnit | None`, so I think just checking `is None` is sufficient here.

**Review by fantix — APPROVED** [review]
Looks good to me!
